src/noise-models.lisp

```diff
@@ -0,0 +1,271 @@
+;;;; noise-models.lisp
+;;;;
+;;;; Author: Sophia Ponte
+
+(in-package #:qvm)
+
+;;; Implementation of rule-based noise models for the QVM.
+
+;;; These values are relatively aribtrary limits for NOISE-PREDICATE
+;;; priorty. They are used to order noise models for matching to
+;;; instructions.
+(defconstant +MAX-PRIORITY+ 20)
+(defconstant +MIN-PRIORITY+ 0)
+(alexandria:define-constant +DEFAULT-NOISE-PREDICATE-NAME+ "NOISE-NAME" :test #'string=)
+
+
+;;; A noise predicate is a function that takes an instruction's gate
+;;; and qubits, and returns either true if the instruction is a match,
+;;; else false. The priority indicates the priority of the predicate
+;;; (tbd) and the noise-position indicates whether the predicate
+;;; should match before or after an instruction.
+(defclass noise-predicate ()
+  ((predicate-function
+    :initarg :predicate-function
+    :reader predicate-function
+    :documentation "Does the instruction trigger the application of a channel?")
+   (priority
+    :initarg :priority
+    :reader priority
+    :documentation "For ordering when checking for an instruction's matching predicates. The convention we use is that more specific predicates get higher priorities.")
+   (noise-position
+    :initarg :noise-position
+    :reader noise-position
+    :type noise-position
+    :documentation "Should application of a channel happen before or after the instruction?")
+   (name
+    :initarg :name
+    :reader name
+    :documentation "Name or descriptor for noise."))
+  (:default-initargs
+   :predicate-function nil
+   :priority 10
+   :noise-position ':after
+   :name +DEFAULT-NOISE-PREDICATE-NAME+)
+  (:documentation "A NOISE-PREDICATE describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATEICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PREDICATE can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
+
+
+(deftype noise-position ()
```

Comment:
I'd move the declaration of this type so that it precedes the declaration of the `noise-predicate` class to help the (human) reader but it's no big deal.